### ISSUE TITLE
[Linux] Graphical artifacts when switching workplaces in i3wm

### ISSUE THREAD

**Issue by FaithLV** [AvaloniaUI/Avalonia#3460]
https://github.com/FaithLV/spectabis-next/issues/5

Using i3wm (i3-gaps).

Launch a GUI window, switch to another workspace and after couple of seconds switch back to the window workspace.

Window will appear fully transparent, artifacted or will display previous application for a few seconds.

Doesn't happen when using Windows 10 workspaces.

Screenshots: https://imgur.com/a/KcS2MGa

**Comment by kekekeks** [comment]
Does i3 use compositing?

**Comment by kekekeks** [comment]
Also try monitoring messages sent to the window on workspace switch (`xwininfo` will give you an id that could be then passed to `xev -id <id>`. If your WM doesn't use compositing, the window should be getting Expose events

**Comment by FaithLV** [comment]
No, i3 by itself doesn't do compositing. I use picom for that. Nevertheless, the issue persists with compositor running or not.

When the window is artifacted, I'm seeing a lot of `FocusOut` events, even thought the window is focused.

**Comment by kekekeks** [comment]
Mkay, we might want to handle `VisibilityUnobscured` and `VisibilityPartiallyObscured` in the same way as `Expose` and trigger a redraw, I guess.

**Comment by Moonlight-Angel** [comment]
Hi, I saw that you implemented what you were talking about in commit https://github.com/AvaloniaUI/Avalonia/commit/ef9bb42252601e50dd55de32b212115f1953474b, so I tried switching to `0.10.0-preview5` to test if it fixed the issue.
Unfortunately, the issue is still present. It looks like it sometimes gets repainted, and sometimes not, but I'm not sure.
I'd want to debug the whole thing and see if it actually asks for a repaint, but I don't know yet how I could do this, so sorry if I can't give any leads for a potential solution. :confused:

**Comment by kekekeks** [comment]
>I'd want to debug the whole thing and see if it actually asks for a repaint, but I don't know yet how I could do this

Running ControlCatalog.NetCore project from a cloned Avalonia solution should just work.

**Comment by Moonlight-Angel** [comment]
Well, that was unexpected. I tried enabling picom (which is a fork of compton) and was pretty sure that would fix the artifacts.
Turns out it doesn't. I was still able to reproduce the issue pretty easily by just going back and forth between two workspaces.
Calling `DoPaint` twice still produces better results overall, but doesn't completely eliminate the artifacts if I really try hard enough.
